### DISCUSSION TITLE
Spike Prime hub's heading() not implemented. Technical limitation?

### DISCUSSION THREAD

**Discussion by ws1088** [pybricks/support#468] [Q&A]
Hi, 
I saw in the beta doc that heading is not implemented in Spike Prime or RI. I do see `yaw_rating()` implemented though. Is there any reason `heading()` is not available? Hardware limitation?
Thanks!

https://beta.pybricks.com/static/docs/hubs/primehub.html?highlight=heading#pybricks.hubs.PrimeHub.imu.heading

**Comment by laurensvalk** [comment]
The sensor itself cannot measure the heading directly. We’d have to calculate it using the measured angular velocity over time. Implementing this requires a fair bit of work and investigation, so we haven’t done that yet.

**Reply by ws1088** [reply]
i see. yeah... even the stock firmware couldn't get the heading calculations right... :(

**Reply by OutOfTheBots** [reply]
I have had some though about the technical difficulty in implementing yaw_heading. 

The obvious difficulty is in the fact the IUM doesn't measure heading but rather measure angular force (how fast the imu is rotating around an axis)  then heading is calculated by knows how fast it is rotating for how long. For any sort of accuracy it requires being calculated at a high enough freq, this is usually done with a hardware timer interrupt.  

This process is quite computationally heavy.

I think it is a good idea that this IUM functionality isn't automatically started when the hub first boots up because many of the programs people create won't be using the IUM. I think it might be better if the was a initalation method that the programmer calls to start the IUM. Something like IUM.init(orientation=axis.X)
